### PR TITLE
app: Add an EOS Discovery Feed Content Provider

### PR DESCRIPTION
**Pull request by smspillaz**
Similar to the included Shell Search Provider, this provider
will include the top five apps to install to populate the
discovery feed with content.

https://phabricator.endlessm.com/T16977

### CODE REVIEW

File: src/gs-discovery-feed-content-provider.c [hunk 1]
```diff
@@ -0,0 +1,296 @@
+/*
+ * gs-shell-search-provider.c - Implementation of an EOS
+ * Discovery Feed Content Provider
+ *
+ * Copyright (C) 2013 Matthias Clasen
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: That should be copyright Endless, I guess.

File: src/gs-discovery-feed-content-provider.c [hunk 2]
```diff
@@ -0,0 +1,296 @@
+/*
+ * gs-shell-search-provider.c - Implementation of an EOS
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: s/gs-shell-search-provider.c/gs-discovery-feed-content-provider.c

File: src/gs-discovery-feed-content-provider.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/*
+ * gs-shell-search-provider.h - Implementation of a GNOME Shell
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.h]: Should be gs-discovery-feed-content-provider.h and change the description.

File: src/gs-discovery-feed-content-provider.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ *   search provider
+ *
+ * Copyright (C) 2013 Matthias Clasen <mclasen@redhat.com>
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.h]: Copyright Endless.

File: src/Makefile.am [hunk 5]
```diff
@@ -260,6 +267,13 @@ gs-shell-search-provider-generated.h gs-shell-search-provider-generated.c: Makef
 		--generate-c-code gs-shell-search-provider-generated \
 		$(srcdir)/shell-search-provider-dbus-interfaces.xml
 
+gs-discovery-feed-content-provider-generated.h gs-discovery-feed-content-provider-generated.c: Makefile.am $(srcdir)/discovery-feed-content-provider-dbus-interfaces.xml
+	$(AM_V_GEN) gdbus-codegen \
+	 	--interface-prefix com.endlessm. \
```
Comment (erikos) [inline src/Makefile.am]: I think there is some erroneous spacing (seen with git diff).

File: src/gs-discovery-feed-content-provider.c [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+static gchar *
+get_app_thumbnail_cached_filename (GsApp *app)
+{
+    const gchar *popular_background_url = gs_app_get_metadata_item (app, "GnomeSoftware::popular-background");
+    g_autofree char *tile_cache_hash = NULL;
+    g_autofree char *cache_identifier = NULL;
+    g_autofree char *url_basename = NULL;
+    char *cache_filename = NULL;
+
+	url_basename = g_path_get_basename (popular_background_url);
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: Indentation.

File: src/gs-discovery-feed-content-provider.c [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+	/* sort by kudos, as there is no ratings data by default */
+	gs_app_list_sort (list, search_sort_by_kudo_cb, NULL);
+
+	g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+	for (i = 0; i < app_list_length; i++) {
+	    gchar *app_thumbnail_filename;
+	    GdkPixbuf *pixbuf;
+	    GVariant *icon_serialized = NULL;
+		GsApp *app = gs_app_list_index (list, i);
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: Indentation.

File: src/gs-discovery-feed-content-provider.c [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+	if (self->cancellable != NULL) {
+		g_cancellable_cancel (self->cancellable);
+		g_clear_object (&self->cancellable);
+	}
+
+	g_application_hold (g_application_get_default ());
+	self->cancellable = g_cancellable_new ();
+	gs_plugin_loader_search_async(self->plugin_loader,
+				       "com.endlessm",
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: Indentation.

File: src/gs-discovery-feed-content-provider.c [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+	}
+	g_dbus_method_invocation_return_value (search->invocation,
+	                                       g_variant_new ("(aa{sv})", &builder));
+
+	pending_search_free (search);
+	g_application_release (g_application_get_default ());
+}
+
+static gboolean
+handle_get_discovery_feed_apps (GsDiscoveryFeedInstallableApps 	   *skeleton,
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: The alignment for the arguments looks a bit inconsistent here. There are a few other places as well.

File: src/gs-discovery-feed-content-provider.c [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+	G_OBJECT_CLASS (gs_discovery_feed_content_provider_parent_class)->dispose (obj);
+}
+
+static void
+gs_discovery_feed_content_provider_init (GsDiscoveryFeedContentProvider *self)
+{
+	self->skeleton = gs_discovery_feed_installable_apps_skeleton_new ();
+
+	g_signal_connect (self->skeleton, "handle-get-relevant-installable-apps",
+			G_CALLBACK (handle_get_discovery_feed_apps), self);
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: Indentation.

File: src/gs-discovery-feed-content-provider.c [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+GsDiscoveryFeedContentProvider *
+gs_discovery_feed_content_provider_new (void)
+{
+	return g_object_new (gs_discovery_feed_content_provider_get_type (), NULL);
+}
+
+void
+gs_discovery_feed_content_provider_setup (GsDiscoveryFeedContentProvider *provider,
+				GsPluginLoader *loader)
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: Alignment.

File: src/gs-discovery-feed-content-provider.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+#define __GS_DISCOVERY_FEED_CONTENT_PROVIDER_H
+
+#include "gs-plugin-loader.h"
+
+#define GS_TYPE_DISCOVERY_FEED_CONTENT_PROVIDER gs_discovery_feed_content_provider_get_type()
+
+G_DECLARE_FINAL_TYPE (GsDiscoveryFeedContentProvider, gs_discovery_feed_content_provider, GS, DISCOVERY_FEED_CONTENT_PROVIDER, GObject)
+
+gboolean		 gs_discovery_feed_content_provider_register	(GsDiscoveryFeedContentProvider	 *self,
+								 GDBusConnection	 *connection,
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.h]: Alignment.

File: src/gs-discovery-feed-content-provider.c [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+	GsDiscoveryFeedContentProvider *self = search->provider;
+	guint i;
+	GVariantBuilder builder;
+	guint app_list_length;
+	guint count;
+	g_autoptr(GsAppList) list = NULL;
+
+	list = gs_plugin_loader_search_finish (self->plugin_loader, res, NULL);
+	if (list == NULL) {
+	    printf("No applications returned, cannot continue");
```
Comment (erikos) [inline src/gs-discovery-feed-content-provider.c]: Debug leftover? :)

**Comment by smspillaz** [comment]
@erikos thanks for the spacing fixup! Looks like I had my editor on the wrong settings. I'll address your other comments now

File: src/Makefile.am (hunk 5 above)
Comment (smspillaz) [inline src/Makefile.am]: Hmm, I don't see anything off here, at least not compared to the shell-search-provider line above

File: src/gs-discovery-feed-content-provider.h (hunk 12 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.h]: Fixed

File: src/gs-discovery-feed-content-provider.c (hunk 2 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: Fixed, thanks for spotting

**Comment by smspillaz** [comment]
@erikos did those fixups and squashed all the commits back into one

File: src/Makefile.am [hunk 14]
```diff
@@ -193,7 +198,9 @@ nodist_gnome_software_SOURCES =				\
 	gs-resources.c					\
 	gs-resources.h					\
 	gs-shell-search-provider-generated.c		\
-	gs-shell-search-provider-generated.h
+	gs-shell-search-provider-generated.h	\
```
Comment (cosimoc) [inline src/Makefile.am]: Nitpick: align the trailing backslash

File: src/discovery-feed-content-provider-dbus-interfaces.xml [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ Lesser General Public License for more details.
+
+ You should have received a copy of the GNU Lesser General
+ Public License along with this library; if not, see <http://www.gnu.org/licenses/>.
+-->
+<node name="/" xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
+  <interface name="com.endlessm.DiscoveryFeedInstallableApps">
+    <method name="GetRelevantInstallableApps">
```
Comment (cosimoc) [inline src/discovery-feed-content-provider-dbus-interfaces.xml]: I think `Relevant` here is a bit redundant, we could just call this `GetInstallableApps()` perhaps.

File: src/gs-application.c [hunk 16]
```diff
@@ -174,7 +176,17 @@ gs_application_dbus_register (GApplication    *application,
 {
 	GsApplication *app = GS_APPLICATION (application);
 	app->search_provider = gs_shell_search_provider_new ();
-	return gs_shell_search_provider_register (app->search_provider, connection, error);
+	app->discovery_feed_content_provider = gs_discovery_feed_content_provider_new ();
+
+	if (!gs_shell_search_provider_register (app->search_provider, connection, error)) {
```
Comment (cosimoc) [inline src/gs-application.c]: I think the coding style here wants no curly braces for one-line if blocks.

File: src/gs-discovery-feed-content-provider.c [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+	g_application_release (g_application_get_default ());
+}
+
+static gboolean
+handle_get_discovery_feed_apps (GsDiscoveryFeedInstallableApps  *skeleton,
+                                GDBusMethodInvocation   *invocation,
+                                gpointer        user_data)
+{
+	GsDiscoveryFeedContentProvider *self = user_data;
+	PendingSearch *pending_search = g_slice_new (PendingSearch);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Use `g_slice_new0()` (or even simply `g_new0()`)

File: src/gs-discovery-feed-content-provider.c [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+	pending_search->invocation = g_object_ref (invocation);
+
+	if (self->cancellable != NULL) {
+		g_cancellable_cancel (self->cancellable);
+		g_clear_object (&self->cancellable);
+	}
+
+	g_application_hold (g_application_get_default ());
+	self->cancellable = g_cancellable_new ();
+	gs_plugin_loader_search_async(self->plugin_loader,
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Missing space before paren

File: src/gs-discovery-feed-content-provider.c [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+                                GDBusMethodInvocation   *invocation,
+                                gpointer        user_data)
+{
+	GsDiscoveryFeedContentProvider *self = user_data;
+	PendingSearch *pending_search = g_slice_new (PendingSearch);
+	pending_search->provider = self;
+	pending_search->invocation = g_object_ref (invocation);
+
+	if (self->cancellable != NULL) {
+		g_cancellable_cancel (self->cancellable);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: This is a bit of a weird pattern; why should another call to the same method cancel the previous one, if still in-flight?
In other words, the cancellable conceptually belongs to the pending search, not the provider instance.

File: src/gs-discovery-feed-content-provider.c [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+	GsDiscoveryFeedContentProvider *self = search->provider;
+	guint i;
+	GVariantBuilder builder;
+	guint app_list_length;
+	guint count;
+	g_autoptr(GsAppList) list = NULL;
+
+	list = gs_plugin_loader_search_finish (self->plugin_loader, res, NULL);
+	if (list == NULL) {
+		g_dbus_method_invocation_return_value (search->invocation, g_variant_new ("(as)", NULL));
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Use the generated `complete_get_relevant_installable_apps()` method here on the skeleton -- that'll avoid using GVariant manually.

File: src/gs-discovery-feed-content-provider.c [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+	g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON (self->skeleton));
+}
+
+static void
+discovery_feed_content_provider_dispose (GObject *obj)
+{
+	GsDiscoveryFeedContentProvider *self = GS_DISCOVERY_FEED_CONTENT_PROVIDER (obj);
+
+	if (self->cancellable != NULL) {
+		g_cancellable_cancel (self->cancellable);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: `g_cancellable_cancel()` works fine with `NULL`

File: src/gs-discovery-feed-content-provider.c [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+
+	app_list_length = gs_app_list_length (list);
+	count = 0;
+
+	/* sort by kudos, as there is no ratings data by default */
+	gs_app_list_sort (list, search_sort_by_kudo_cb, NULL);
+
+	g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+	for (i = 0; i < app_list_length; i++) {
+	        gchar *app_thumbnail_filename;
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: You can use `g_autofree` for this

File: src/gs-discovery-feed-content-provider.c [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+	        GdkPixbuf *pixbuf;
+	        GVariant *icon_serialized = NULL;
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: You can use `g_autoptr` for this

File: src/gs-discovery-feed-content-provider.c [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+	g_autofree char *tile_cache_hash = NULL;
+	g_autofree char *cache_identifier = NULL;
+	g_autofree char *url_basename = NULL;
+	char *cache_filename = NULL;
+
+	url_basename = g_path_get_basename (popular_background_url);
+	tile_cache_hash = g_compute_checksum_for_string (G_CHECKSUM_SHA256,
+	                                                 popular_background_url,
+	                                                 -1);
+	cache_identifier = g_strdup_printf("%s-%s", tile_cache_hash, url_basename);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Missing space before paren

File: src/gs-discovery-feed-content-provider.c [hunk 25]
```diff
@@ -0,0 +1,295 @@
...
+	cache_filename = gs_utils_get_cache_filename ("eos-popular-app-thumbnails",
+	                                              cache_identifier,
+	                                              GS_UTILS_CACHE_FLAG_NONE,
+	                                              NULL);
+
+	/* Check to see if the file exists in the cache at the time we called this
+	 * function. If it does, then we can use this app in the results
+	 * since it will have a thumbnail */
+	if (g_file_test (cache_filename, G_FILE_TEST_EXISTS)) {
+		g_debug("Hit cache for thumbnail when loading discovery feed %s: %s", popular_background_url, cache_filename);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Missing space before paren

File: src/gs-discovery-feed-content-provider.c [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+		return cache_filename;
+	}
+
+    return cache_filename;
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: I think here you meant to free `cache_filename` and return `NULL`? By the way `cache_filename` can also use `g_autofree()` (and `g_steal_pointer()`)

File: src/gs-discovery-feed-content-provider.c [hunk 27]
```diff
@@ -0,0 +1,295 @@
...
+	guint weight;
+} DiscoveryFeedKudoWeight ;
+
+static const DiscoveryFeedKudoWeight discovery_feed_app_kudos[] =
+{
+	{ GS_APP_KUDO_MY_LANGUAGE, 10 },
+	{ GS_APP_KUDO_FEATURED_RECOMMENDED, 5 },
+	{ GS_APP_KUDO_POPULAR, 3 }
+};
+static const guint discovery_feed_app_kudos_len = sizeof (discovery_feed_app_kudos) / sizeof (DiscoveryFeedKudoWeight);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: You can just use `G_N_ELEMENTS()` instead of this

Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: Ah yes, that's the macro I was looking for.

File: src/gs-discovery-feed-content-provider.c (hunk 24 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: Oops  - I should fix this in the place I copied it from too.

File: src/gs-discovery-feed-content-provider.c (hunk 22 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: I'm not sure if we can actually - it is a loop variable and I'm pretty sure that `g_autofree` only happens on function exit and not scope exit.

File: src/gs-discovery-feed-content-provider.c (hunk 23 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: See above, I'm not sure about that.

File: src/gs-discovery-feed-content-provider.c (hunk 19 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: I've copied this from the search-provider code, so you'd have to refer to that  😎. But I think your analysis makes sense here, we can just remove this.

File: src/gs-discovery-feed-content-provider.c (hunk 26 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: I think in both cases, the intention is to transfer ownership to the caller. However, I noticed that we don't free memory on the caller side, so I should do that.

File: src/gs-discovery-feed-content-provider.c (hunk 21 above)
Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: I decided to just remove the cancellable

File: src/gs-discovery-feed-content-provider.c [hunk 28]
```diff
@@ -0,0 +1,287 @@
...
+
+	/* Check to see if the file exists in the cache at the time we called this
+	 * function. If it does, then we can use this app in the results
+	 * since it will have a thumbnail */
+	if (g_file_test (cache_filename, G_FILE_TEST_EXISTS)) {
+		g_debug ("Hit cache for thumbnail when loading discovery feed %s: %s", popular_background_url, cache_filename);
+		return cache_filename;
+	}
+
+	return cache_filename;
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: I still don't understand why you're not returning `NULL` here in this case. Either that's incorrect, or both the comment above and the `NULL` check in the caller are wrong.

File: src/gs-discovery-feed-content-provider.c [hunk 29]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+static const guint N_APPS_TO_INCLUDE_IN_RESULT = 5;
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: I'd love to move this kind of filtering to the client side; is that possible?

Comment (smspillaz) [inline src/gs-discovery-feed-content-provider.c]: I mean, we can, its just that searching + refining hundreds of apps on every query could be a little expensive.

File: src/gs-discovery-feed-content-provider.c [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+	                                              cache_identifier,
+	                                              GS_UTILS_CACHE_FLAG_NONE,
+	                                              NULL);
+
+	/* Check to see if the file exists in the cache at the time we called this
+	 * function. If it does, then we can use this app in the results
+	 * since it will have a thumbnail */
+	if (g_file_test (cache_filename, G_FILE_TEST_EXISTS)) {
+		g_debug ("Hit cache for thumbnail when loading discovery feed %s: %s", popular_background_url, cache_filename);
+		return g_steal_pointer(&cache_filename);
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Missing space before paren

File: src/gs-discovery-feed-content-provider.c [hunk 31]
```diff
@@ -0,0 +1,282 @@
...
+
+static guint
+get_discovery_feed_app_kudo_score (GsApp *app)
+{
+	guint i = 0;
+	guint score = 0;
+	guint64 kudos = gs_app_get_kudos (app);
+
+	for (; i < discovery_feed_app_kudos_len; ++i) {
+		if ((kudos & discovery_feed_app_kudos[i].kudo) != 0) {
```
Comment (cosimoc) [inline src/gs-discovery-feed-content-provider.c]: Remove curly braces for single-line statement

**Pull request merged**